File: 0056-merge-intervals/0056-merge-intervals.cpp
// TC- O(nlogn) where n is number of intervals
// SC - O(logn) for sorting
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {

        // sort intervals because if not, then O(n^2). 
        sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b){return a[0]<b[0];});

        vector<vector<int>> ans;
        for(int i=0; i<intervals.size(); i++){
            int min = intervals[i][0];
            int max = intervals[i][1];
            while(i+1<intervals.size() and intervals[i+1][0]<=max){
                if(intervals[i+1][1]>max)
                    max = intervals[i+1][1];
                i++;
            }
            ans.push_back({min, max});
        }
        return ans;
    }
};

File: 0056-merge-intervals/0056-merge-intervals.py
# TC- O(nlogn)
# SC- O(n)

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        ans = []
        i = 0
        intervals.sort()
        while i<len(intervals):
            start = intervals[i][0]
            end = intervals[i][1]
            while(i+1<len(intervals) and intervals[i+1][0]<=end):
                end = max(end, intervals[i+1][1])
                i += 1
            ans.append([start, end])
            i += 1

        return ans